File: game.py
import sys
import random
import hashlib
import hmac

class Game:
    def __init__(self, moves):
        self.moves = moves
        self.rules = self.generate_rules()
        self.key = self.generate_key()
        self.computer_move = random.choice(moves)

    def generate_key(self):
        return hashlib.sha256(bytes(random.getrandbits(8) for _ in range(32))).digest()

    def generate_rules(self):
        num_moves = len(self.moves)
        middle = num_moves // 2
        rules = {}
        for i, move in enumerate(self.moves):
            winning_moves = [self.moves[(i + j) % num_moves] for j in range(1, middle + 1)]
            losing_moves = [self.moves[(i - j) % num_moves] for j in range(1, middle + 1)]
            rules[move] = {'win': winning_moves, 'lose': losing_moves}
        return rules

    def generate_hmac(self, move):
        return hmac.new(self.key, move.encode(), hashlib.sha256).hexdigest()

    def determine_winner(self, user_move):
        if user_move == self.computer_move:
            return "Draw"
        if user_move in self.rules[self.computer_move]['win']:
            return "You win!"
        return "Computer wins."

    def play(self):
        print("Available moves:")
        for i, move in enumerate(self.moves):
            print(f"{i + 1} - {move}")
        print("0 - Exit")
        
        while True:
            try:
                choice = int(input("Enter your move: "))
                if choice == 0:
                    print("Exiting the game.")
                    return
                elif 1 <= choice <= len(self.moves):
                    user_move = self.moves[choice - 1]
                    hmac_value = self.generate_hmac(user_move)
                    
                    print(f"Your move: {user_move}")
                    print(f"Computer's move: {self.computer_move}")
                    print(f"HMAC key: {self.key.hex()}")
                    print(f"HMAC: {hmac_value}")
                    
                    result = self.determine_winner(user_move)
                    print(f"Result: {result}")
                    return
                else:
                    print("Invalid input. Please enter a valid choice.")
            except ValueError:
                print("Invalid input. Please enter a valid choice.")

if __name__ == "__main__":
    if len(sys.argv) < 4 or len(sys.argv) % 2 != 1:
        print("Usage: python game.py move1 move2 move3 ...")
        sys.exit(1)

    moves = sys.argv[1:]
    game = Game(moves)
    game.play()

File: game.js
const crypto = require('crypto');

class MovesTable {
    constructor(moves) {
        this.moves = moves;
        this.table = this.generateTable();
    }

    generateTable() {
        const table = [['', ...this.moves]];
        for (const move of this.moves) {
            const row = [move];
            for (const otherMove of this.moves) {
                if (this.isWinningMove(move, otherMove)) {
                    row.push('Win');
                } else if (move === otherMove) {
                    row.push('Draw');
                } else {
                    row.push('Lose');
                }
            }
            table.push(row);
        }
        return table;
    }

    isWinningMove(move, otherMove) {
        const half = Math.floor(this.moves.length / 2);
        const startIndex = this.moves.indexOf(move);
        const endIndex = (startIndex + half) % this.moves.length;
        return this.moves.slice(startIndex + 1, startIndex + half + 1).includes(otherMove)
            || this.moves.slice(0, endIndex + 1).includes(otherMove);
    }

    printTable() {
        for (const row of this.table) {
            console.log(row.join('\t'));
        }
    }
}

class Game {
    constructor(moves) {
        this.moves = moves;
        this.movesTable = new MovesTable(this.moves);
        this.key = crypto.randomBytes(32);
    }

    generateMove() {
        return this.moves[Math.floor(Math.random() * this.moves.length)];
    }

    computeHmac(move) {
        const hmac = crypto.createHmac('sha256', this.key);
        hmac.update(move);
        return hmac.digest('hex');
    }

    play() {
        const computerMove = this.generateMove();
        console.log(`Computer's move: ${computerMove}`);

        const userMove = this.getUserMove();
        const hmacValue = this.computeHmac(userMove);
        console.log(`HMAC: ${hmacValue}`);

        const winner = this.movesTable.table[this.moves.indexOf(userMove) + 1][this.moves.indexOf(computerMove) + 1];
        console.log(`Winner: ${winner}`);
        console.log(`Key: ${this.key.toString('hex')}`);
    }

    getUserMove() {
        const readline = require('readline-sync');
        while (true) {
            console.log('Choose your move:');
            for (let i = 0; i < this.moves.length; i++) {
                console.log(`${i + 1} - ${this.moves[i]}`);
            }
            const choice = readline.questionInt();
            if (choice >= 1 && choice <= this.moves.length) {
                return this.moves[choice - 1];
            } else {
                console.log('Invalid choice. Please try again.');
            }
        }
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 3 || args.length % 2 !== 1 || new Set(args).size !== args.length) {
        console.log('Invalid input. Please provide an odd number of unique moves.');
        return;
    }

    const game = new Game(args);
    game.movesTable.printTable();
    game.play();
}

main();
